### ISSUE TITLE
Result of `ckks/encoder/(ecd *encoderComplex128) Encode` is not decoded properly when encoded with a `Scale` exactly `2.0**64` 

### ISSUE THREAD

**Issue by MichaelA-CR** [tuneinsight/lattigo#395]
Hi Lattigo's team,

There is an bug with the `CKKSEncode` function, when asked to encode with a scale of exactly in the interval `[2.0**64 - 2**10; 2.0**64]`.
I've attached a standalone [main.go.txt](https://github.com/tuneinsight/lattigo/files/12301447/main.go.txt)  file to reproduce the issue. It encodes a vector full of `1`'s, and decodes it right after. It then displays the precision of the decoded value.

Here are the outputs:
```bash
CKKS parameters: logN=16, logSlots=15, H(192; 32), sigma=3.200000, logQP=1520, levels=26, scale=2^40.000000

Precision of values vs. ciphertext with Scale = 18446744073709551616.000000

Level: 25 (logQ = 1276)
Scale: 18446744073709551616.000000
ValuesTest: (0.5000000000+0.0000000000i) (0.5000000000+0.0000000000i) (0.5000000000+0.0000000000i) (0.5000000000+0.0000000000i)...
ValuesWant: (1.0000000000+0.0000000000i) (1.0000000000+0.0000000000i) (1.0000000000+0.0000000000i) (1.0000000000+0.0000000000i)...

┌─────────┬───────┬───────┬───────┐
│    Log2 │ REAL  │ IMAG  │ L2    │
├─────────┼───────┼───────┼───────┤
│MIN Prec │  1.00 │  +Inf │  1.00 │
│MAX Prec │  1.00 │  +Inf │  1.00 │
│AVG Prec │  1.00 │  +Inf │  1.00 │
│MED Prec │  1.00 │  +Inf │  1.00 │
└─────────┴───────┴───────┴───────┘
Err STD Slots  : 38.00 Log2
Err STD Coeffs : 31.00 Log2



Precision of values vs. ciphertext with Scale = 36893488147419103232.000000

Level: 25 (logQ = 1276)
Scale: 36893488147419103232.000000
ValuesTest: (1.0000000000+0.0000000000i) (1.0000000000+0.0000000000i) (1.0000000000+0.0000000000i) (1.0000000000+0.0000000000i)...
ValuesWant: (1.0000000000+0.0000000000i) (1.0000000000+0.0000000000i) (1.0000000000+0.0000000000i) (1.0000000000+0.0000000000i)...

┌─────────┬───────┬───────┬───────┐
│    Log2 │ REAL  │ IMAG  │ L2    │
├─────────┼───────┼───────┼───────┤
│MIN Prec │  +Inf │  +Inf │  +Inf │
│MAX Prec │  +Inf │  +Inf │  +Inf │
│AVG Prec │  +Inf │  +Inf │  +Inf │
│MED Prec │  +Inf │  +Inf │  +Inf │
└─────────┴───────┴───────┴───────┘
Err STD Slots  :  -Inf Log2
Err STD Coeffs :  -Inf Log2
```

We already identified the culprit code, and the fix seems to be easy. In `ckks/utils/SingleFloatToFixedPointCRT` a conversion between a `float` to `uint64` causes it. This could be fixed by the following modification:
```go
diff --git a/ckks/utils.go b/ckks/utils.go
index b7341660..e09ffd1d 100644
--- a/ckks/utils.go
+++ b/ckks/utils.go
@@ -187,7 +187,7 @@ func SingleFloatToFixedPointCRT(r *ring.Ring, i int, value float64, scale float6

        moduli := r.ModuliChain()[:r.Level()+1]

-       if value > 1.8446744073709552e+19 {
+       if value >= 1.8446744073709552e+19 {
                xFlo = big.NewFloat(value)
                xFlo.Add(xFlo, big.NewFloat(0.5))
                xInt = new(big.Int)
```

Thanks for your work.

**Comment by Pro7ech** [comment]
Hi @MichaelA-CR, thank you for reporting this bug. It is indeed occurring when `values * scale = 2^{64}` which causes an `uint64` overflow. And changing `>` to `>=` indeed fixes the problem. I have added the fix to the 4.2 release.

**Comment by Pro7ech** [comment]
#379
